### ISSUE TITLE
[no-redeclare] Spurious warning on a double export of const+type

### ISSUE THREAD

**Issue by Zamiell** [typescript-eslint/typescript-eslint#4904]
## Repro

For some use-cases, you can't use "normal" TypeScript enums, so you revert to using `const`. When creating enums like this "by hand", you need to export the type in addition to the enum itself, like this:

#### `Fruit.ts`
```ts
export const Fruit = {
  Apple: 0,
  Banana: 1,
  Pear: 2,
} as const;
export type Fruit = typeof Fruit[keyof typeof Fruit]
```

#### `someFile.ts`
```ts
import { Fruit } from "./Fruit";

function useFruit(fruit: Fruit) {}

useFruit(Fruit.Apple);
```

In this case, the `no-redeclare` rule is triggered.

To my knowledge, this is a common-ish pattern in TypeScript. For example, the [`protobuf-ts`](https://github.com/timostamm/protobuf-ts) tool creates TypeScript code that double-exports both constants and types.

## Solution

`@typescript-eslint/eslint-plugin` simply extends the base ESLint rule of `no-redeclare`.

I think the TypeScript-version of the rule should detect this specific situation and not throw an error.

**Comment by bradzacher** [comment]
This is explicitly called out in the docs

https://typescript-eslint.io/rules/no-redeclare#ignoredeclarationmerge

It is intentional because 99.99% of the time you do not want this behaviour as it creates a confusing DevX and codebase to have a variable and a type named the same.

---

> For some use-cases, you can't use "normal" TypeScript enums, so you revert to using const. 

What are the use cases wherein you can't use an enum? A TS enum is literally just sugar for the example code you added so I don't understand why an enum wouldn't work.

**Comment by Zamiell** [comment]
Oh, sorry about the noise, I'll close this issue.

**Comment by Zamiell** [comment]
> What are the use cases wherein you can't use an enum?

When you have to set the enum value to a constructor function, like:

```ts
enum Fruit {
  Apple = getFruitValue(0),
  Banana = getFruitValue(1),
}
```

**Comment by Zamiell** [comment]
> This is explicitly called out in the docs

Hrm, even with the `ignoreDeclarationMerge` option turned on, the rule still warns about the code that I have listed in the OP.

> It is intentional because 99.99% of the time you do not want this behaviour as it creates a confusing DevX and codebase to have a variable and a type named the same.

I was under the impression that it makes the development experience better.

In other words, enums work like this:

#### `Fruit.ts`
```ts
export enum Fruit {
  Apple,
  Banana,
}
```
#### `main.ts`
```ts
function useFruit(fruit: Fruit) {}

useFruit(Fruit.Apple);
```

If I change the code in the OP to be exported uniquely, then it would result in the following code:

#### `main.ts`
```ts
function useFruit(fruit: FruitType) {}

useFruit(Fruit.Apple);
```

As a developer, I might be left scratching my head and as to why FruitType is different from Fruit.

**Comment by Zamiell** [comment]
> Yes - as mentioned it's called out in the docs.

Ah, sorry, I see now, my mistake!

> This is a very, very rare usecase. In fact I've never seen someone create a pseudo-enum like this.

Interesting!

Ok, thanks, very good points here. Yes, API leaking is an issue. :(

In my case, I want to provide a similar API for "real" enums alongside "fake" enums that have values that absolutely require a constructor.

So I'll have to carefully balance the pros and cons here!
